subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -1312,10 +1319,12 @@ static void lwm2m_rd_client_service(struct k_work *work)
 
 		case ENGINE_BOOTSTRAP_REG_SENT:
 			/* wait for bootstrap registration done */
+			timeout = EXCHANGE_LIFETIME;
 			break;
 
 		case ENGINE_BOOTSTRAP_REG_DONE:
 			/* wait for transfer done */
+			timeout = EXCHANGE_LIFETIME;
```

Comment:
I see the problem with `ENGINE_BOOTSTRAP_REG_DONE` as in this particular case there is no timeout implemented, as we rely on server-initiated communication (Bootstrap-Finish). But do we need to introduce a separate timeout mechanism for other states? We already have timeout callbacks called when there is no response after serveral retransmission attempts (CoAP-driven):
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_rd_client.c#L442

Reply:
Rest of the timeouts are based on CoAP layer signalling us.

But there is not timeout if there is no CoAP transaction ongoing. Therefore I think the only way to introduce timeouts is to implement this kind of watch-dog mechanism that is not supposed to trigger when all is working correctly.

Reply:
Hmm, I had an impression that being in one of the `_SENT` states implies that there is an ongoing transaction, but maybe there are some cases I don't see. Anyway, having such how you nicely called it "watch-dog" at the state machine level shouldn't hurt anyway.